libs/shell/shell.c

```diff
@@ -0,0 +1,209 @@
+#include <stdio.h>
+#include <string.h>
+
+#include <FreeRTOS.h>
+#include <task.h>
+#include <queue.h>
+#include <semphr.h>
+
+#include <stdbool.h>
+#include <util.h>
+#include <stdio.h>
+#include <shell.h>
+
+char *shell_prompt_text = "> ";
+char *shell_banner = "\n\nCommand Shell";
+
+struct shell_state {
+	/* +1 for the help command */
+	struct shell_cmd commands[SHELL_MAX_COMMANDS + 1];
+	int command_count;
+	char input_buf[32];
+	int input_pos;
+	bool initialized;
+	xSemaphoreHandle shell_lock;
+} shell_state;
+
+static bool shell_is_whitespace(char c)
+{
+	switch(c) {
+	case ' ':
+	case '\t':
+	case '\r':
+	case '\n':
+		return true;
+	default:
+		break;
+	}
+	return false;
+}
+
+static void shell_exec_cmd(int argc, char **argv)
+{
+	unsigned int cmd_len;
+	unsigned int tgt_cmd_len = strlen(argv[0]);
+	int i;
+	for (i = 0; i < shell_state.command_count; i++) {
+		cmd_len = strlen(shell_state.commands[i].command);
+
+		if (cmd_len == tgt_cmd_len && shell_state.commands[i].action &&
+		    strncmp(argv[0], shell_state.commands[i].command, tgt_cmd_len) == 0) {
+			shell_state.commands[i].action(argc, argv);
+			putchar('\n');
+			return;
+		}
+	}
+	printf("Invalid Command!\n");
+}
+
+static void shell_parse_command(void)
+{
+	int i;
+	int arg_count = 0;
+	char *arg_list[8] = {NULL};
+	bool in_arg = false;
+
+	/* TODO: Fix the bug where I pass everything as args...*/
+	for(i = 0; i < shell_state.input_pos; i++) {
+		if (shell_is_whitespace(shell_state.input_buf[i]) && arg_count == 0)
+			continue;
+
+		if (shell_is_whitespace(shell_state.input_buf[i])) {
+			if (in_arg) {
+				shell_state.input_buf[i] = 0x00;
+				in_arg = false;
+			}
+		} else if (!in_arg) {
+			arg_list[arg_count] = &shell_state.input_buf[i];
+			arg_count++;
+			in_arg = true;
+		}
+	}
+	shell_state.input_buf[i] = 0x00;
+
+	if (arg_count)
+		shell_exec_cmd(arg_count, arg_list);
+}
+
+static int shell_process_input(void)
+{
+	char c;
+
+	c = (char)getchar();
+
+	switch(c) {
+	case '\r': /*Fall through*/
+	case '\n':
+		putchar('\n');
+		/* handle command here */
+		shell_parse_command();
```

Comment:
If you ever get "hello world\r\n" as input, then you're going to process it as two commands ("hello world" and ""). Easy way to handle this case is to only call shell_parse_command() if input_pos != 0.